src/client/entity-editor/relationship-editor/relationship-editor.js

```diff
@@ -0,0 +1,339 @@
+	onClose?: () => mixed,
+	onSave?: (_Relationship) => mixed
+};
+
+type RelationshipModalState = {
+	entityType?: ?EntityTypeObject,
+	relationshipType?: ?RelationshipType,
+	relationship?: ?_Relationship,
+	targetEntity?: ?EntitySearchResult
+};
+
+function getInitState(
+	baseEntity: Entity, initRelationship: ?_Relationship
+): RelationshipModalState {
+	if (_.isNull(initRelationship)) {
+		return {
+			entityType: null,
+			relationship: null,
+			relationshipType: null,
+			targetEntity: null
+		};
+	}
+
+	const sourceEntity = _.get(initRelationship, ['sourceEntity']);
+	const targetEntity = _.get(initRelationship, ['targetEntity']);
+
+	const baseEntityBBID = _.get(baseEntity, ['bbid']);
+
+	const sourceEntityBBID = _.get(sourceEntity, ['bbid']);
+	const otherEntity = baseEntityBBID === sourceEntityBBID ?
+		targetEntity : sourceEntity;
+
+	const searchFormatOtherEntity = otherEntity && {
+		id: _.get(otherEntity, ['bbid']),
+		text: _.get(
+			otherEntity, ['defaultAlias', 'name'], '(unnamed)'
+		),
+		type: _.get(otherEntity, ['type']),
+		value: _.get(otherEntity, ['bbid'])
+	};
+
+	return {
+		entityType: {
+			id: _.lowerCase(_.get(searchFormatOtherEntity, 'type')),
+			label: _.get(searchFormatOtherEntity, 'type')
+		},
+		relationship: initRelationship,
+		relationshipType: _.get(initRelationship, ['relationshipType']),
+		targetEntity: searchFormatOtherEntity
+	};
+}
+
+class RelationshipModal
+	extends React.Component<RelationshipModalProps, RelationshipModalState> {
+	static defaultProps = {
+		onCancel: null,
+		onClose: null,
+		onSave: null
+	};
+
+	constructor(props: RelationshipModalProps) {
+		super(props);
+
+		this.state = getInitState(props.baseEntity, props.initRelationship);
+	}
+
+	handleEntityTypeChange = (value: EntityTypeObject) => {
+		this.setState({
+			entityType: value,
+			relationship: null,
+			relationshipType: null,
+			targetEntity: null
+		});
+	};
+
+	handleEntityChange = (value: EntitySearchResult) => {
+		this.setState({
+			relationship: null,
+			relationshipType: null,
+			targetEntity: value
+		});
+	};
+
+	handleRelationshipTypeChange = (value: _Relationship) => {
+		this.setState({
+			relationship: value,
+			relationshipType: value.relationshipType
+		});
+	};
+
+	handleSave = () => {
+		const {onSave} = this.props;
+		if (onSave) {
+			if (this.state.relationship) {
+				onSave(this.state.relationship);
+			}
+		}
+	};
+
+	calculateProgressAmount() {
+		if (!this.state.entityType) {
+			return 1;
+		}
+
+		if (!this.state.targetEntity) {
+			return 30;
+		}
+
+		if (!this.state.relationshipType) {
+			return 60;
+		}
+
+		return 100;
+	}
+
+	renderEntityTypeSelect() {
+		return (
+			<FormGroup>
+				<ControlLabel>Entity Type</ControlLabel>
+				<ReactSelect
+					name="entityType"
+					options={ENTITY_TYPE_VALUES}
+					value={this.state.entityType}
+					onChange={this.handleEntityTypeChange}
+				/>
+			</FormGroup>
+		);
+	}
+
+	renderEntitySelect() {
+		return (
+			<EntitySearchField
```

Comment:
The EntitySearchField is limited to 10 results. Is that on purpose?

Reply:
It's the default limit for Elasticsearch - see https://www.elastic.co/guide/en/elasticsearch/reference/5.5/search-request-from-size.html

We can make this bigger, but maybe better after this PR.